### PR TITLE
#13: Extract overall coverage from Jacoco report

### LINKED ISSUE #13: Fix parsing of Jacoco report
A Jacoco report file contains coverage information in multiple levels of XML, e.g.

```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<!DOCTYPE report PUBLIC "-//JACOCO//DTD Report 1.1//EN" "report.dtd">
<report name="Quality Summarizer Maven Plugin">
    <sessioninfo id="hostname" start="1729610403849" dump="1729610408295" />
    <package name="com/exasol/swqa/maven">
        <class name="com/exasol/swqa/maven/SummarizerMojo" sourcefilename="SummarizerMojo.java">
            <method name="&lt;init&gt;" desc="(Lorg/apache/maven/project/MavenProject;)V" line="43">
                <counter type="INSTRUCTION" missed="0" covered="6" />
                <counter type="LINE" missed="0" covered="3" />
                <counter type="COMPLEXITY" missed="0" covered="1" />
                <counter type="METHOD" missed="0" covered="1" />
            </method>
            <method name="execute" desc="()V" line="51">
                <counter type="INSTRUCTION" missed="0" covered="11" />
                <counter type="LINE" missed="0" covered="5" />
                <counter type="COMPLEXITY" missed="0" covered="1" />
                <counter type="METHOD" missed="0" covered="1" />
            </method>
            <!-- ... -->
            <counter type="INSTRUCTION" missed="20" covered="237" />
            <counter type="BRANCH" missed="2" covered="6" />
            <counter type="LINE" missed="5" covered="55" />
            <counter type="COMPLEXITY" missed="2" covered="14" />
            <counter type="METHOD" missed="0" covered="12" />
            <counter type="CLASS" missed="0" covered="1" />
        </class>
        <sourcefile name="SummarizerMojo.java">
            <line nr="43" mi="0" ci="2" mb="0" cb="0" />
            <line nr="44" mi="0" ci="3" mb="0" cb="0" />
            <!-- ... -->
            <counter type="INSTRUCTION" missed="20" covered="237" />
            <counter type="BRANCH" missed="2" covered="6" />
            <counter type="LINE" missed="5" covered="55" />
            <counter type="COMPLEXITY" missed="2" covered="14" />
            <counter type="METHOD" missed="0" covered="12" />
            <counter type="CLASS" missed="0" covered="1" />
        </sourcefile>
        <counter type="INSTRUCTION" missed="20" covered="237" />
        <counter type="BRANCH" missed="2" covered="6" />
        <counter type="LINE" missed="5" covered="55" />
        <counter type="COMPLEXITY" missed="2" covered="14" />
        <counter type="METHOD" missed="0" covered="12" />
        <counter type="CLASS" missed="0" covered="1" />
    </package>
    <counter type="INSTRUCTION" missed="20" covered="237" />
    <counter type="BRANCH" missed="2" covered="6" />
    <counter type="LINE" missed="5" covered="55" />
    <counter type="COMPLEXITY" missed="2" covered="14" />
    <counter type="METHOD" missed="0" covered="12" />
    <counter type="CLASS" missed="0" covered="1" />
</report>
```

Currently, the plugin takes the first `counter` element with type `BRANCH` (`<counter type="BRANCH" ... />`) and uses its coverage information. This way, the plugin will only use the coverage for a single method, not for the complete project.

Instead, we need to use the top level element `/report/counter[type="BRANCH"]`.

### PR DESCRIPTION
**Pull request by kaklakariada**
Closes #13

### CODE REVIEW

File: doc/changes/changes_0.2.0.md [hunk 1]
```diff
@@ -1,15 +1,16 @@
-# Quality Summarizer Maven Plugin 0.2.0, released 2024-??-??
+# Quality Summarizer Maven Plugin 0.2.0, released 2024-10-23
 
 Code name: Downgrade Maven Version
 
 ## Summary
 
-This version downgrades the required Maven version from 3.8.8 to 3.8.7, allowing this plugin to work on GitHub action runners `windows-latest`. The release also allows reading jacoco report with doctype declaration.
+This version downgrades the required Maven version from 3.8.8 to 3.8.7, allowing this plugin to work on GitHub action runners `windows-latest`. The release also allows reading jacoco report with doctype declaration and extracts the overall coverage from the Jacoco report.
```
Comment (redcatbear) [inline doc/changes/changes_0.2.0.md]:
```suggestion
This version downgrades the required Maven version from 3.8.8 to 3.8.7, allowing this plugin to work on GitHub action runners `windows-latest`. The release also allows reading JaCoCo report with `doctype` declaration and extracts the overall coverage from the JaCoCo report.
```

File: doc/changes/changes_0.2.0.md [hunk 2]
```diff
@@ -1,15 +1,16 @@
...
 
 ## Bugfixes
 
 * #9: Downgrade required Maven version
 * #7: Allow reading jacoco report with doctype declaration.
+* #13: Extract overall coverage Jacoco report
```
Comment (redcatbear) [inline doc/changes/changes_0.2.0.md]:
```suggestion
* #7: Allow reading JaCoCo report with `doctype` declaration
* #13: Extract overall coverage JaCoCo report
```

File: src/test/java/com/exasol/swqa/maven/SummarizerMojoTest.java [hunk 3]
```diff
@@ -56,7 +48,67 @@ void extractCoverageFromXmlVersionTag() throws MojoFailureException, IOException
...
+                    <counter type="BRANCH" missed="0" covered="100" /> <!-- This is the element we need -->
+                    <counter type="LINE" missed="5" covered="55" />
+                    <counter type="COMPLEXITY" missed="2" covered="14" />
+                    <counter type="METHOD" missed="1" covered="12" />
+                    <counter type="CLASS" missed="1" covered="1" />
+                </report>
+                """), equalTo(100.0F));
+    }
+
+    private String jacocoReport(final String xmlPrefix, final int missed, final int covered) {
```
Comment (redcatbear) [inline src/test/java/com/exasol/swqa/maven/SummarizerMojoTest.java:102]: You made the prefix configurable, but it is the same for all test cases.

**Review by redcatbear — CHANGES_REQUESTED** [review]

File: src/test/java/com/exasol/swqa/maven/SummarizerMojoTest.java (hunk 3 above)
Comment (kaklakariada) [reply src/test/java/com/exasol/swqa/maven/SummarizerMojoTest.java:102]: Some tests are not shown in the diff. The following prefixes are used:

```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<!DOCTYPE report PUBLIC "-//JACOCO//DTD Report 1.1//EN" "report.dtd">
```

```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```

and an empty string `""`

**Review by redcatbear — APPROVED** [review]

**Pull request merged**